File: apps/api/app/schemas/ranked.py
"""경쟁전 REST 응답 모델 (PROTOCOL §1.2).

`messages.py` 와 나눠 둔 이유는 길이뿐이다(파일당 400줄).

**숨은 MMR 은 어떤 모델에도 없다.** 그 숫자가 보이는 순간 사람들은 티어가 아니라
그걸 보고 놀게 되고, "왜 이겼는데 RR 이 조금 올랐냐"가 계산식 폭로전이 된다.
"""

from __future__ import annotations

import datetime as dt

from pydantic import BaseModel, Field


class TierResponse(BaseModel):
    """티어 한 칸의 표시 정보. 프런트가 뱃지를 그릴 때 쓴다."""

    index: int
    group: str
    group_name: str
    #: 1~3. 디비전이 없는 계급(레디언트)은 0.
    division: int
    name: str
    color: str


class SeasonResponse(BaseModel):
    id: int
    key: str
    name: str
    started_at: dt.datetime
    ended_at: dt.datetime | None = None
    is_active: bool


class RankStats(BaseModel):
    """한 시즌의 랭크 상태. 배치 중이면 `placed=False` 이고 tier/rr 은 의미가 없다."""

    #: 1~25. 0 이면 아직 배치 중이다.
    tier: int = 0
    rr: int = 0
    placements: int = 0
    placement_total: int = 5
    placed: bool = False
    peak_tier: int = 0
    peak_rr: int = 0
    wins: int = 0
    losses: int = 0
    #: 양수면 연승, 음수면 연패.
    streak: int = 0
    best_streak: int = 0
    rounds_won: int = 0
    rounds_lost: int = 0


class MyRankResponse(BaseModel):
    """`GET /api/ranked/me`. 아직 한 판도 안 했으면 rank 는 전부 기본값이다."""

    season: SeasonResponse
    rank: RankStats
    #: 리더보드 순위. 배치 중이면 None.
    position: int | None = None


class LeaderboardEntry(BaseModel):
    position: int
    account_id: str
    nickname: str
    customization: dict = Field(default_factory=dict)
    rank: RankStats


class LeaderboardResponse(BaseModel):
    season: SeasonResponse
    entries: list[LeaderboardEntry] = Field(default_factory=list)


class MatchHistoryEntry(BaseModel):
    """전적 한 줄. 경쟁전이 아니면 rr_delta 는 0 이다."""

    id: str
    ranked: bool
    mode: str
    map_id: str
    rounds: int
    duration_sec: int
    #: 상대가 도중에 나가서 끝난 판.
    forfeit: bool
    ended_at: dt.datetime
    won: bool
    score: int
    opponent_score: int
    opponent_nickname: str
    rr_delta: int = 0
    #: 배치전 중이었던 판. RR 대신 "배치 n/5" 로 보여 준다.
    placement: bool = False
    tier_before: int = 0
    tier_after: int = 0
    rr_after: int = 0


class MatchHistoryResponse(BaseModel):
    entries: list[MatchHistoryEntry] = Field(default_factory=list)

File: apps/web/src/types/ranked.ts
// 경쟁전 타입 — docs/PROTOCOL.md §1.2 와 1:1 대응.
//
// **티어 표를 여기 적어 두지 않는다.** 이름과 색은 서버(`GET /api/ranked/tiers`)가
// 유일한 원본이고, 프런트는 부팅할 때 한 번 받아서 캐시한다(`@/api/ranked`).
// 같은 표를 양쪽에 적어 두면 계급을 하나 손대는 순간 반드시 어긋난다.

/** 티어 한 칸. index 는 1(아이언 1) ~ 25(레디언트), 0 은 미배치. */
export interface TierInfo {
  index: number;
  /** 'iron' | 'bronze' | … | 'radiant'. 뱃지 모양을 이 값으로 가른다. */
  group: string;
  group_name: string;
  /** 1~3. 디비전이 없는 계급(레디언트)은 0. */
  division: number;
  name: string;
  color: string;
}

export interface Season {
  id: number;
  key: string;
  name: string;
  started_at: string;
  ended_at: string | null;
  is_active: boolean;
}

/** 한 시즌의 내 랭크. placed 가 false 면 tier/rr 은 아직 의미가 없다. */
export interface RankStats {
  tier: number;
  rr: number;
  placements: number;
  placement_total: number;
  placed: boolean;
  peak_tier: number;
  peak_rr: number;
  wins: number;
  losses: number;
  /** 양수면 연승, 음수면 연패. */
  streak: number;
  best_streak: number;
  rounds_won: number;
  rounds_lost: number;
}

export interface MyRank {
  season: Season;
  rank: RankStats;
  /** 리더보드 순위. 배치 중이면 null. */
  position: number | null;
}

export interface LeaderboardEntry {
  position: number;
  account_id: string;
  nickname: string;
  customization: Record<string, unknown>;
  rank: RankStats;
}

export interface LeaderboardPage {
  season: Season;
  entries: LeaderboardEntry[];
}

/** 전적 한 줄. 일반전이면 rr_delta 는 0 이다. */
export interface MatchRecord {
  id: string;
  ranked: boolean;
  mode: string;
  map_id: string;
  rounds: number;
  duration_sec: number;
  /** 상대가 도중에 나가서 끝난 판. */
  forfeit: boolean;
  ended_at: string;
  won: boolean;
  score: number;
  opponent_score: number;
  opponent_nickname: string;
  rr_delta: number;
  placement: boolean;
  tier_before: number;
  tier_after: number;
  rr_after: number;
}

/**
 * 매치가 끝난 뒤 서버가 밀어 주는 랭크 변동(WS `rank_update`).
 *
 * DB 쓰기가 끝난 다음에 오므로 `match_over` 이벤트보다 **한 박자 늦다** —
 * 종료 오버레이는 이 값이 아직 없을 때도 그려질 수 있어야 한다.
 */
export interface RankChange {
  won: boolean;
  rr_delta: number;
  /** 배치전이 아직 진행 중이라 티어가 없다. */
  placement: boolean;
  placement_played: number;
  placement_total: number;
  /** 이 판으로 배치가 끝나 티어가 처음 정해졌다. */
  placed_now: boolean;
  promoted: boolean;
  demoted: boolean;
  tier_before: number;
  tier_after: number;
  rr_before: number;
  rr_after: number;
  /** 변동 후 티어의 표시 정보(서버가 같이 실어 준다). */
  tier: TierInfo;
}
